### PR TITLE
Account autocreation from LDAP after reverse proxy authentication

### PR DESCRIPTION
**Pull request by pboguslawski**
Gitea allows autocreation of account from external source after successful
basic auth but not after successful reverse proxy auth. This mod adds such
feature.

Related: https://github.com/gogs/gogs/issues/2498
Author-Change-Id: IB#1104925

### CODE REVIEW

**Comment by pboguslawski** [comment]
BTW: please consider refactoring gitea code to separate authentication logic from authorization logic and from user data synchronization logic and moving all auth/authz/sync configuration from SQL to app.ini; mixing these things makes changes harder and error prone; configuration in SQL is a pain when using automated provisioning; also external passwords like LDAP are problably less secure when stored in SQL (and every SQL backup...).

BTW2: please consider synchronization of user avatar from jpegPhoto LDAP attribute (if exists).

BTW3: please consider disallowing local (in gitea) user attributes modifications for attributes imported from LDAP (i.e. first name, surname, e-mail, is active, is admin, etc.); LDAP synchronization often (always?) means that admin wants to manage this data in LDAP not in app.

**Comment by lafriks** [comment]
> BTW: please consider refactoring gitea code to separate authentication logic from authorization logic and from user data synchronization logic and moving all auth/authz/sync configuration from SQL to app.ini; mixing these things makes changes harder and error prone; configuration in SQL is a pain when using automated provisioning; also external passwords like LDAP are problably less secure when stored in SQL (and every SQL backup...).

I don't agree as that loses ability to configure from UI and secrets in DB is maybe even more secure than in config file.

> 
> BTW2: please consider synchronization of user avatar from jpegPhoto LDAP attribute (if exists).

There is already issue for this. Avatar should probably be synchronized only on login not to have too much load on sync process. 

> 
> BTW3: please consider disallowing local (in gitea) user attributes modifications for attributes imported from LDAP (i.e. first name, surname, e-mail, is active, is admin, etc.); LDAP synchronization often (always?) means that admin wants to manage this data in LDAP not in app.

Please submit separate issue for this.

**Comment by pboguslawski** [comment]
> please consider refactoring gitea code to separate authentication logic from authorization logic and from user data synchronization logic and moving all auth/authz/sync configuration from SQL to app.ini

Current design does not allow application admin to be denied access for messing low level configuration like LDAP backends (this should be server admin not application admin business).

> Please submit separate issue for this.

Please take a look at https://github.com/go-gitea/gitea/pull/13068

**Comment by pboguslawski** [comment]
Just fixed another problem with reverse proxy login - user lang was ignored and csrf error was fired on first user settings saving. Login with header use now session to avoid setting cookies with every request; session is skipped for API calls like in sspi_windows.go.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
agh wrong button!

File: modules/auth/ldap/ldap.go [hunk 1]
```diff
@@ -230,12 +230,22 @@ func checkRestricted(l *ldap.Conn, ls *Source, userDN string) bool {
...
+	if directBind && alreadyAuthenticated {
+		log.Debug("Cannot bind using user %s credentials - user already authenticated. BindDN must be used.", name)
+		return nil
+	}
+
+	if !ls.AttributesInBind && alreadyAuthenticated {
+		log.Debug("Cannot get attributes using user %s credentials - user already authenticated; --attributes-in-bind must be used.", name)
+		return nil
+	}
+
```
Comment (zeripath) [inline modules/auth/ldap/ldap.go:248]:
```suggestion
	if alreadyAuthenticated {
		if directBind {
			log.Debug("Cannot bind pre-authenticated user %s. BindDN must be used.", name)
			return nil
		}

		if !ls.AttributesInBind {
			log.Debug("Cannot get attributes for pre-authenticated user %s without --attributes-in-bind.", name)
			return nil
		}
	} else if len(passwd) == 0 {
		log.Debug("Auth. failed for %s, password cannot be empty", name)
		return nil
	}
```

File: models/login_source.go [hunk 2]
```diff
@@ -731,8 +731,8 @@ func ExternalUserLogin(user *User, login, password string, source *LoginSource)
 	return user, nil
 }
 
-// UserSignIn validates user name and password.
-func UserSignIn(username, password string) (*User, error) {
+// UserSignIn validates user name and password. Password verification in LDAP skipped if already authenticated.
+func UserSignIn(username, password string, alreadyAuthenticated bool) (*User, error) {
```
Comment (zeripath) [inline models/login_source.go:735]: I guess one thing I don't really like about this is that the user has already signed in - we're just abusing this function to lookup the user in the external source - and only in one place (at least so far). Maybe we should be doing a ExternalUserLookup or something like that.

Comment (zeripath) [reply models/login_source.go:735]: Further if the user is in the db then this alreadyAuthenticated path fails.

Comment (pboguslawski) [reply models/login_source.go:735]: > Further if the user is in the db then this alreadyAuthenticated path fails.

This mod works ok for us in scenario with LDAP user backend + reverse proxy auth using HTTP header. No problems with initial account creation nor logging in/updating existing accounts from LDAP. If you see any bug please describe how to reproduce.

> Maybe we should be doing a ExternalUserLookup or something like that.

This mod was created for our needs with minimal changes in upstream code. Feel free to enhance it.

**Comment by 6543** [comment]
@pboguslawski what is the current status if this?

**Comment by pboguslawski** [comment]
This mod was created for our needs with minimal changes in upstream code (which should be refactored as above IHMO) and works ok for us. Feel free to enhance it or close this PR if you don't like it.

**Comment by zeripath** [comment]
I guess this is going to miss 1.15 - My suspicion is that we need to think about authentication, authorization and user registration slightly differently. Currently registration is very tightly bound to the authentication source that it is associated with.

I'm gonna move this to 1.16 but reference #16199 to remind myself.

**Comment by pboguslawski** [comment]
Replaced by https://github.com/go-gitea/gitea/pull/18452

**Pull request closed without merging**